sandboxes/heka/output/heka_s3_partition.lua

```diff
@@ -0,0 +1,281 @@
+                    error(string.format("field '%s' allowed_values range must have a 'min' or 'max'", name))
+                end
+                if av.min and type(av.min) ~= "string"
+                or av.max and type(av.max) ~= "string" then
+                    error(string.format("field '%s' allowed_values range min/max must be a string", name))
+                end
+                d.matcher = function (v) return is_in_range(v, av.min, av.max) end
+            end
+        else
+            error(string.format("field '%s' allowed_values invalid type: %s", type(av)))
+        end
+    end
+    return df.dimensions
+end
+
+
+local function get_fqfn(path)
+    return string.format("%s/%s", batch_path, path)
+end
+
+
+local function close_fh(entry)
+    if not entry[2] then return end
+    entry[2]:close()
+    entry[2] = nil
+    fh_cnt = fh_cnt - 1
+end
+
+
+local function copy_file(path, entry)
+    close_fh(entry)
+    local t = os.time()
+    local cmd
+    if t == time_t then
+        buffer_cnt = buffer_cnt + 1
+    else
+        time_t = t
+        buffer_cnt = 0
+    end
+
+    local src  = get_fqfn(path)
+    cmd = string.format("aws s3 cp %s %s/%s/%d_%d_%s", src, s3_path,
+                        string.gsub(path, "+", "/"), time_t, buffer_cnt, hostname)
+
+    local ret = os.execute(cmd)
+    if ret ~= 0 then
+        return string.format("ret: %d, cmd: %s", ret, cmd)
+    end
+    files[path] = nil
+
+    local ok, err = os.remove(src);
+    if not ok then
+        return string.format("os.remove('%s') failed: %s", path, err)
+    end
+end
+
+
+local function get_entry(path)
+    local ct = os.time()
+    local t = files[path]
+    if not t then
+        t = {ct, nil} -- last active, file handle
+        files[path] = t
+    else
+        t[1] = ct
+    end
+
+    if not t[2] then
+        if max_file_handles ~= 0 then
+            if fh_cnt >= max_file_handles then
+                local oldest = ct + 60
+                local entry
+                for k,v in pairs(files) do -- if we max out file handles a lot we will want to make this more efficient
+                    local et = v[1]
+                    if v[2] and et < oldest then
+                        entry = v
+                        oldest = et
+                    end
+                end
+                if entry then close_fh(entry) end
+            end
+        end
+        t[2] = assert(io.open(get_fqfn(path), "a"))
+        fh_cnt = fh_cnt + 1
+    end
+    return t
+end
+
+local dimensions = validate_dimensions(read_config("dimension_file"))
+
+function process_message()
+    local dims = {}
+    for i,d in ipairs(dimensions) do
+        local v = sanitize_dimension(read_message(d.field_name))
+        if v then
+            if d.matcher(v) then
+                dims[i] = v
+            else
+                dims[i] = "OTHER"
+            end
+        else
+            dims[i] = "UNKNOWN"
+        end
+    end
+    local path = table.concat(dims, "+")
```

Comment:
I suggest adding a comment here and in the `copy_file` function describing that the `+` will be converted to a `/` on S3.